rfcs/text/0003_application_service_mounting.md

```diff
@@ -0,0 +1,259 @@
+- Start Date: 2019-05-10
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+A front-end service to manage registration and root-level routing for
+first-class applications.
+
+# Basic example
+
+```tsx
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import { MyApp } from './components';
+
+class MyPlugin {
+  setup({ application }) {
+    application.registerApp({
+      id: 'my-app',
+      title: 'My Application',
+      rootRoute: '/myapp',
```

Comment:
Maybe `rootRoute` could be just `id` prefixed with slash?

Reply:
FWIW I prefer the explicitness of requiring it (`rootPath` could be another name without the mouthful of root route, but also who cares)

Reply:
+1 for `...Path` instead of `...Route` in my experience most frameworks define a "route" as the combination of a "path" and the business logic to handle it. Since we already have a configurable `basePath` this isn't the root or the base path, but in reality a second-level path. What do you think about calling this `mountPath`?

Reply:
One of the benefits of not supporting rootRoute and instead using `/app/<id>` is that the server _could_ do some light data pre loading (like infected vars) based I the app ID in the url.

Reply:
I'm opting to remove this option. Unless there's a strong usecase for applications that have `id`s that do not match their routes, this is superfluous.